### PR TITLE
Change Events#ssl_error signature from (error, peeraddr, peercert) to (error, ssl_socket)

### LINKED ISSUE #2337: Requests fail during phased restarts, hot restarts, and shutdown
**Describe the bug**

During a phased restart, some clients do not get successful responses from the server.

Requests fail in a variety of ways. The ones I've observed and that I'm able to reproduce reliably are these:

1. The server closes the connection without first writing a response onto the socket `curl: (52) Empty reply from server`
2. `curl: (56) Recv failure: Connection reset by peer`
3. The exception `Unable to add work while shutting down` is raised, invoking the `lowlevel_error_handler` (by default, this returns a response with response code 500)

**Puma config:**

```ruby
port 3000
workers 2
```

```sh
PUMA_MASTER_PIDFILE=/usr/src/app/puma.pid
bundle exec puma --pidfile "${PUMA_MASTER_PIDFILE}"
```

**To Reproduce**

I've created a repo that reliably reproduces the errors: https://github.com/cjlarose/puma-phased-restart-errors

**Expected behavior**

All requests are passed to the application and the application's responses are all delivered to the requesting clients. The [restart documentation for puma][0] suggests that phased restarts are meant to have zero downtime.

[0]: https://github.com/puma/puma/blob/a24b51b294ff8dd2511f910abe1a0db0d66ed43a/docs/restart.md

**Desktop (please complete the following information):**
 - OS: Linux
 - Puma Version: 4.3.5

### PR DESCRIPTION
**Pull request by MSP-Greg**
### Description

The method signature should include the socket, so it determines what socket info is logged, and also so it can catch exceptions.

Closes #2365, closes #2367

### Your checklist for this pull request
- [x] I have reviewed the [guidelines for contributing](../blob/master/CONTRIBUTING.md) to this repository.
- [x] I have added an entry to [History.md](../blob/master/History.md) if this PR fixes a bug or adds a feature. If it doesn't need an entry to HISTORY.md, I have added `[changelog skip]` or `[ci skip]` to the pull request title.
- [x] I have added appropriate tests if this PR fixes a bug or adds a feature.
- [x] My pull request is 100 lines added/removed or less so that it can be easily reviewed.
- [ ] If this PR doesn't need tests (docs change), I added `[ci skip]` to the title of the PR.
- [x] If this closes any issues, I have added "Closes `#issue`" to the PR description or my commit messages.
- [x] I have updated the documentation accordingly.
- [x] All new and existing tests passed, including Rubocop.

### CODE REVIEW

File: lib/puma/events.rb [hunk 1]
```diff
@@ -106,10 +106,12 @@ def parse_error(error, req)
...
-    # and +peercert+ any peer certificate (if present).
+    # @param error <Puma::MiniSSL::SSLError>
+    # @param ssl_socket <Puma::MiniSSL::Socket>
     #
-    def ssl_error(error, peeraddr, peercert)
+    def ssl_error(error, ssl_socket)
+      addr = ssl_socket.peeraddr.last rescue "<unknown>"
+      peercert = ssl_socket.peercert
       subject = peercert ? peercert.subject : nil
       @error_logger.info(error: error, text: "SSL error, peer: #{peeraddr}, peer cert: #{subject}")
```
Comment (nateberkopec) [inline lib/puma/events.rb:116]: peeraddr is undefined. did you mean to change this line to `peer: #{addr}`?

**Review by nateberkopec — CHANGES_REQUESTED** [review]

File: lib/puma/events.rb (hunk 1 above)
Comment (MSP-Greg) [reply lib/puma/events.rb:116]: You're being too nice.  Actually, I'm changing `addr` on line 113 to `peeraddr`.  More importantly, I'm adding a test (bad assumption on my part).  Give me a minute...

Comment (MSP-Greg) [reply lib/puma/events.rb:116]: Fixed and added a test.  Thanks.

**Comment by nateberkopec** [comment]
Nice to see @ocowchun on the co-author there, nice touch. Co-authoring is underused!

**Comment by ocowchun** [comment]
Haha, Thanks, but I don't think I deserve the credit. Still, I wish I can have more contribution to puma in the future.

**Pull request merged**